### PR TITLE
[WIP]add postgresql plugin agent  sql param show

### PR DESCRIPTION
**Pull request by aderm**
Please answer these questions before submitting pull request

- Why submit this pull request?
- [ ] Bug fix
- [X] New feature provided
- [ ] Improve performance

- Related issues

___
### Bug fix
- Bug description.

- How to fix?

___
### New feature or improvement
- Describe the details and related test reports.

### CODE REVIEW

**Comment by wu-sheng** [comment]
I didn't see rebase :) Still a lot of commits.

**Comment by wu-sheng** [comment]
You should learn from the existing case(https://github.com/apache/skywalking/tree/master/test/plugin/scenarios/solrj-7.x-scenario), and move the following two into the main repo.
1. https://github.com/SkyAPMTest/agent-auto-integration-testcases/tree/master/postgresql-scenario
1. https://github.com/SkyAPMTest/agent-auto-integration-testcases/tree/master/postgresql-9.4-scenario

You could ping me on QQ, wechat or mail, as we don't have the document(we are working on that) for the new plugin test development guide.

**Comment by wu-sheng** [comment]
The old test includes 55 cases, https://github.com/SkyAPMTest/agent-integration-test-report#postgresql. Last time, we passed. This should be easy work to move them in.

**Comment by aderm** [comment]
@wu-sheng ok,thanks

**Comment by wu-sheng** [comment]
Any update about the postgre test PR?

**Comment by aderm** [comment]
WIP

File: apm-sniffer/apm-sdk-plugin/jdbc-commons/src/main/java/org/apache/skywalking/apm/plugin/jdbc/define/StatementEnhanceInfos.java [hunk 1]
```diff
@@ -73,6 +73,22 @@ public void setParameter(int index, final Object parameter) {
...
 
+    public String getFullSql() {
+        StringBuilder resultSql = new StringBuilder();
+        int index = 0;
+        int startPos = 0;
+        int findPos = 0;
+        while ((findPos = sql.indexOf("?", startPos)) > 0) {
+            resultSql.append(sql.substring(startPos, findPos));
+            resultSql.append("'");
+            resultSql.append(parameters[index++]);
```
Comment (kezhenxu94) [inline apm-sniffer/apm-sdk-plugin/jdbc-commons/src/main/java/org/apache/skywalking/apm/plugin/jdbc/define/StatementEnhanceInfos.java]: It's unsafe to parse the SQL using `?`, are there possibilities when the sql includes `?` itself in the value, not a parameter placeholder?

`select * from news_table where author=? and title like '%?'`, I know it's a little tricky, just to remind the risk here

Comment (kezhenxu94) [reply apm-sniffer/apm-sdk-plugin/jdbc-commons/src/main/java/org/apache/skywalking/apm/plugin/jdbc/define/StatementEnhanceInfos.java]: My point is not to try to parse the SQL, store the parameters in another tag

Comment (wu-sheng) [reply apm-sniffer/apm-sdk-plugin/jdbc-commons/src/main/java/org/apache/skywalking/apm/plugin/jdbc/define/StatementEnhanceInfos.java]: Parsing SQL is the next generation after ShardingShpere ( @tristaZero and  @terrymanu ) provides the real SQL Parser.  We don't do this by ourselves.

**Comment by wu-sheng** [comment]
Please fix the conflict. And we could continue this now.

**Comment by wu-sheng** [comment]
This should continue.

**Comment by aderm** [comment]
ok

File: apm-sniffer/apm-sdk-plugin/jdbc-commons/src/main/java/org/apache/skywalking/apm/plugin/jdbc/define/StatementEnhanceInfos.java (hunk 1 above)
Comment (aderm) [reply apm-sniffer/apm-sdk-plugin/jdbc-commons/src/main/java/org/apache/skywalking/apm/plugin/jdbc/define/StatementEnhanceInfos.java]: @wu-sheng @kezhenxu94 ACK，i will store the parameters in another tag.

File: apm-sniffer/apm-agent-core/src/main/java/org/apache/skywalking/apm/agent/core/conf/Config.java [hunk 2]
```diff
@@ -273,6 +273,22 @@
             public static int SQL_PARAMETERS_MAX_LENGTH = 512;
         }
 
+        public static class PostgreSQL {
+            /**
+             * If set to true, the parameters of the sql (typically {@link java.sql.PreparedStatement}) would be
+             * collected.
+             */
+            public static boolean TRACE_SQL_PARAMETERS = true;
```
[the remark this answers is not in the document]
Comment (aderm) [reply apm-sniffer/apm-agent-core/src/main/java/org/apache/skywalking/apm/agent/core/conf/Config.java]: How much impact on performance？

**Comment by wu-sheng** [comment]
Please resolve the conflict.

**Comment by aderm** [comment]
refer #3695

**Comment by wu-sheng** [comment]
Should close this one?

**Comment by kezhenxu94** [comment]
replaced by #3695

**Pull request closed without merging**